### PR TITLE
Improve authentication procedure

### PR DESCRIPTION
**Pull request by JohnnyJayJay**
The aim of this PR is to make ReGreet more suitable for the range of authentication procedures greetd supports. In its current state, ReGreet only really focuses on one style of authentication: username - password. However some setups do not match this, for example:

- fingerprint authentication (via fprint) works, but does not feel natural, as you are not logged in as soon as you've authenticated. Instead, you have to press login again. This current behaviour is mostly coincidental.
- If you use authentication that requires input but the input prompt doesn't contain the word "password" (possibly because you use a different language, a different type of secret input, or a visible input) you cannot use ReGreet because it doesn't show an input field
- Multi-step or fallback authentication is not supported, ReGreet will error or restart the session respectively.

To fix this and support all of greetd's authentication types, I have revised the way greetd responses are handled, taking inspiration from [agreety](https://github.com/kennylevinsen/greetd/blob/23fadf16d444164983784db5ce9d91d2c655fcd6/agreety/src/main.rs#L77). In case of an `info` or `error` authentication message, ReGreet now sends an empty authentication response immediately and waits for the next greetd response. Similarly, after an input such as a password is submitted, the exact same handling is performed, meaning that if greetd asks for more authentication, ReGreet can handle it now regardless of type.

This is the main change in this PR. A couple other changes have been made:
- ReGreet displays the authentication message from greetd as the input label now, instead of always "Password:". In the simple case of basic password authentication, the result is the same.
- `password_mode` has been replaced by `input_mode`, a three state enum specifying whether there is currently a secret input (equivalent to password_mode = true), a visible input (new feature) or no input (equivalent to password_mode = false).
- Most occurrences of "password" have been replaced by "input" or "secret", depending on whether the previous use was general or password-authentication specific. This is in line with greetd naming.

I've marked this PR as a draft because I still want to change the way responses are handled in a loop - currently this seems to block gtk/relm updates, meaning that info/error messages are not displayed. Maybe this can be circumvented by making use of `sender.oneshot_command(...)` or similar. I am practically completely unfamiliar with gtk and relm, so I'm open for suggestions.

Generally, I'll be happy to answer any questions you might have about the changes and discuss revisions.

### CODE REVIEW

File: src/gui/component.rs
Comment (rharish101) [inline src/gui/component.rs:17]: This could be simplified to `use super::model::InputMode`.

File: src/gui/component.rs
Comment (rharish101) [inline src/gui/component.rs:278]: I'm not so sure about this. Have you tested it?

IMHO `model.updates.input` should change the text of both `secret_entry` and `visible_entry`, but not the other way around (as far as my understanding of Relm4 goes). If my hunch is correct, then you'll have to use a conditional on `model.updates.input_mode()` to use either `secret_entry` or `visible_entry`.

File: src/gui/model.rs
Comment (rharish101) [inline src/gui/model.rs:330]: Looking at the code, I agree with you. If you tested this, and it works, then feel free to go ahead and remove the input reset.

File: src/gui/model.rs
Comment (rharish101) [inline src/gui/model.rs:192]: Could you clarify how the input (such as a fingerprint) is sent to greetd here? Does it automatically get it from the fingerprint drivers? If yes, would the loop be running infinitely until a fingerprint is given? If so, why not add a waiting time, so as not to unnecessarily keep looping too fast?

Also, what does an empty auth message signify?

**Review by rharish101 — COMMENTED** [review]
Thanks a lot for the PR!

I don't have a fingerprint scanner set up, or know of other auth methods. Is it possible for me to test this locally somehow (since we don't have any tests set up for ReGreet yet)?

**Comment by rharish101** [comment]
> I've marked this PR as a draft because I still want to change the way responses are handled in a loop - currently this seems to block gtk/relm updates, meaning that info/error messages are not displayed. Maybe this can be circumvented by making use of sender.oneshot_command(...) or similar.

I'm new to GTK and Relm too, so I'm not sure how to resolve this. We would need to see if the code and the GUI actually run on separate threads. If not, then maybe it's not possible to make GUI updates without yielding from the loop.

In general, I'm not clear about why exactly we need to be looping (other than that's what agreety does), so I can't offer any ideas at the moment.

File: src/gui/component.rs
Comment (JohnnyJayJay) [reply src/gui/component.rs:276]: It does work - I tested it by changing `secret_entry` to `visible_entry` in this piece of code and using password login.

I agree that this is hacky and I actually wanted to implement your idea first - branching based on `input_mode`. However I wasn't sure how to do that given that this is in a closure that moves the variables. If you know how to access the model in the closure I'll definitely update it, otherwise I could maybe ask somewhere else for advice.

File: src/gui/model.rs
Comment (JohnnyJayJay) [reply src/gui/model.rs:192]: My information comes mostly from greetd-ipc(7) and inferred details from agreety. So, heads up, I didn't look through the greetd source myself (yet?), but this makes pretty good sense to me and so far it's been confirmed in my experiments.

So, this is how it works when using something like `fprint` in conjunction with greetd, from what I gather.

greetd tries an authentication step and therefore speaks indirectly to fprint. It gets a message like "Place your index finger on the fingerprint reader" and fprint then just waits for the driver, by itself. Now, greetd will give you this initial information via an authentication message of type info. If you now satisfy the request and make it read your fingerprint, authentication will be done but the greeter won't be finished because greetd doesn't send you a message without you sending one first. You'll have to send an authentication response to finish the process (that is why you have to press login twice currently).

The manpages say that an authentication response is always required:

> Indicates that an authentication message needs to be answered to con‐
tinue through the authentication flow. There are no limits on the num‐
ber and type of messages that may be required for authentication to 
succeed, and a greeter should not make any assumptions about the mes‐
ages. Must be answered with either post_auth_message_response or cancel_session.

So, when you respond with an empty authentication response immediately after receiving the info, greetd will wait for the authentication and you actually won't get a response until there is feedback from fprint. Thus, there actually is no CPU-intensive loop here because greetd makes you wait.

In other words, you could say that sending an empty authentication response is like reading a line of output from fprint/PAM, if that makes sense?

In short:
1. ReGreet -> greetd (create session)
2. greetd -> PAM -> fprint (authenticate)
3. greetd -> ReGreet (initial output from fprint)
4. Subsequent requests to greetd with empty authentication responses wait for fprint output, i.e. further info, failure or success
5. You use the scanner: fprint -> greetd (success/failure)
6. After the next (or currently waiting) request to greetd: greetd -> ReGreet (authentication success/error)

> Also, what does an empty message signify?
Hopefully my elaboration already cleared it up somehow, but I'll also quote greetd-ipc(7) again on empty responses (although it's not much of an explanation):

> If the message was informative (info, error), then a response does not need to be set in this message.

**Comment by JohnnyJayJay** [comment]
Thanks for your quick review.

> I don't have a fingerprint scanner set up, or know of other auth methods. Is it possible for me to test this locally somehow (since we don't have any tests set up for ReGreet yet)?

Afaik there are multiple procedures that work like fingerprints. For example, smart cards/security keys should work the same (as in, greetd gets a response from the device directly, rather than from greeter input). So if you have one of those, you could set PAM up with it and test. I have a Yubikey, so I think I will actually try this as well for good measure.

In theory it's also possible through other (software-based) means, but I haven't looked into that. Like you could have a PAM setup that calls some custom software that simply waits for a keystroke or similar... Maybe something like that already exists? If it doesn't, it should only take a couple lines of code to make something like that yourself. I only know the basics of PAM though, so I'd have to do some research myself. [Pamtester](https://linux.die.net/man/1/pamtester) could come in handy to make something like that.

> We would need to see if the code and the GUI actually run on separate threads. If not, then maybe it's not possible to make GUI updates without yielding from the loop.

In this case, the code is called from the thread that receives the login button press event, so I'm assuming it's the gtk thread. This also explains why UI elements become unresponsive in that case.
If there is a way to tell GTK to "force" an update, that could be another way to resolve it.

> In general, I'm not clear about why exactly we need to be looping

See my response to one of your comments. It doesn't have to be a literal loop, but in case of an info or error message there should be an immediate response.

**Comment by fufexan** [comment]
Hello, just chiming in to say facial authentication through [Howdy](https://github.com/boltgolt/howdy/) works the same way in ReGreet, but it has an option that "virtually presses login again" to authenticate directly after confirming the face of the user.
Would this behaviour collide with this PR's functionality?

**Comment by JohnnyJayJay** [comment]
> Hello, just chiming in to say facial authentication through [Howdy](https://github.com/boltgolt/howdy/) works the same way as fprintd in ReGreet, but it has an option that "virtually presses login again" to authenticate directly after confirming the face of the user.
> Would this behaviour collide with this PR's functionality?

I'm not exactly sure what you mean by "virtually presses login again", since the authentication doesn't run inside ReGreet. Could you elaborate? My initial intuition would be that this PR can't really conflict with anything because it "just" causes ReGreet to get updates from greetd immediately.

Thanks for mentioning this though, I might check out whether it works myself.

File: src/gui/model.rs
Comment (rharish101) [reply src/gui/model.rs:192]: Ah, got it. Thanks a lot for the detailed explanation!

File: src/gui/component.rs
Comment (rharish101) [reply src/gui/component.rs]: I found out how to do this. I simply made the `InputMode` enum `Copy`, so that it can be copied inside the closure automatically.

**Comment by fufexan** [comment]
@JohnnyJayJay It's [this option](https://github.com/boltgolt/howdy/blob/beta/howdy/src/config.ini#L33-L37) in howdy's config.

**Comment by rharish101** [comment]
> Maybe this can be circumvented by making use of sender.oneshot_command(...) or similar.

We can try this out and see if it works. You would need to add a new command message variant to `CommandMsg`: https://github.com/rharish101/ReGreet/blob/95175466f5bd44d2e188fc3486d819c209173211/src/gui/messages.rs#L64-L69

Then, you'd have to handle this in `update_cmd_with_view` here: https://github.com/rharish101/ReGreet/blob/95175466f5bd44d2e188fc3486d819c209173211/src/gui/component.rs#L341-L347

After these, you'll have to send a command to the sender in the greetd auth loop using `sender.spawn_command`. In my experience, `(spawn_)oneshot_command` never worked for me, which is why I use `sender.spawn_command` throughout the codebase. Maybe this has changed with the latest Relm4 revisions, because I experienced this in v0.5.0-rc.1

**Comment by JohnnyJayJay** [comment]
> I might check out whether it works myself.

So... installing howdy currently doesn't work for me, I'll have to do this another time.

> After these, you'll have to send a command to the sender in the greetd auth loop using sender.spawn_command

Some changes will have to be made in order for us to be able to use the variables in the command function. Not sure what'll be the best way of going about that. In any case, I asked in the relm4 Matrix channel for advice, perhaps there is a way out of this that doesn't involve restructuring the app.

**Comment by JohnnyJayJay** [comment]
So, I talked with someone on Matrix and they suggested that the preferred option would indeed be to work with commands/threads/async. However there is a also another way that does not seem terrible in this case and lets us preserve the simple nature of the program. Apparently you can use `Component::update_with_view` to register a hook that updates the view every time a certain input is received. So we could do that for a special kind of input, and that input is then sent in every iteration of the loop.
Need to look into it more, but that seems to be the basic idea.

**Comment by rharish101** [comment]
Ah, so you can directly update the view without updating the model? Sounds good to me.

**Comment by JohnnyJayJay** [comment]
Cheers, I'm back from vacation.

After further exchange with people on Matrix, I've come to the conclusion that `update_with_view` is not an option for us after all. The way I see it now (after finally understanding the required relm concepts properly) there are 2 options:
- Use a oneshot command in `handle_greetd_response` that sends a message to greetd on another thread and gives back the response as a `CommandOutput`, which then in turn is handled in `update_cmd` by calling `handle_greetd_response` again. As far as I can tell, this requires 2 changes: `UserSessInfo` should be made part of the model instead of passing it to functions directly. I think this is more in line with the new request-response dynamic anyway. Furthermore, we need to be able to share the greetd client among multiple threads. I'm somewhat of a Rust noob, but I know that there are multiple reference/pointer types at our disposal to enable this.
- Use [async](https://relm4.org/book/stable/async.html) components. From what I've been told this shouldn't actually change too much about the component definitions and be less work than I thought initially.

Since I've never used Rust async before and I already have a plan for 1., I'm trying that first. 2. might be a good future goal.

**Comment by JohnnyJayJay** [comment]
My last change seems to work. However it's not displaying "Logging in..." now anymore after it's done. Will fix that later. I'm also working on a separate label for error messages, so both info and error can be displayed at the same time.

**Comment by JohnnyJayJay** [comment]
Somewhere in the last commits there has been a regression that messes up the focus of the secret input field (presumably visible too). It keeps re-focusing/selecting the most recently entered characters which makes it impossible to enter anything. I initially thought this was because both visible and secret input grab focus if any input mode is active, but changing this didn't fix it. As far as I can tell, my change to make the time updates more relm-idiomatic didn't cause it either. I'm going back through the commits now and hope it's not some weird async thing...

File: src/gui/model.rs
Comment (JohnnyJayJay) [reply src/gui/model.rs:341]: Decided to keep it like this now that I changed the structure again a bit. And resetting the input does fit here, so I'll just leave it alone.

**Comment by JohnnyJayJay** [comment]
Changed some former ReGreet behaviour: Session gets cancelled now when greetd sends an authentication error. Need feedback on this change - does greetd allow you to resubmit a password if you got an error in the same session? For me this doesn't seem to work, but I'm not sure if it's supposed to. In case it works for other people I'll revert this.

I'll mark this PR as ready for review now. The only thing left that I know I want to do is style the new error label. Maybe an error symbol, red colour, things like that. But idk anything about styling stuff, so I'm open for suggestions/edits.

**Comment by rharish101** [comment]
Once again, thanks for the PR and your time. I'll review it soon, but I'm gonna need a bit of time.

> Since I've never used Rust async before and I already have a plan for 1., I'm trying that first. 2. might be a good future goal.

Sounds good. I already have some experience working with async in Rust, so I'll keep option 2 in mind.

> Sidenote: wow, this project seems to have become a lot more popular 😄

Yeah, I advertised this on [r/unixporn](https://www.reddit.com/r/unixporn/comments/11bixwm/greetd_oc_i_made_a_clean_and_customizable_login/), which explains the rise in popularity.

> The only thing left that I know I want to do is style the new error label. Maybe an error symbol, red colour, things like that. But idk anything about styling stuff, so I'm open for suggestions/edits.

I can take a look at it and find a way to style it/fit it in 👍🏼.

> I'll try to whip something up that lets people without "unusual" authentication methods at their disposal test the changes. I will also take another look at facial recognition with howdy if I can get it to work.

That would be amazing.

File: src/gui/model.rs
Comment (rharish101) [inline src/gui/model.rs:282]: This could be simplified with an `if let` instead of a `match`.

**Review by rharish101 — CHANGES_REQUESTED** [review]
LGTM, with some minor nitpicks. I'll make the edits for the error message and merge this, although I'd prefer to somehow test this (or have at least one other user test this).

**Review by rharish101 — APPROVED** [review]

**Comment by JohnnyJayJay** [comment]
> LGTM, with some minor nitpicks. I'll make the edits for the error message and merge this, although I'd prefer to somehow test this (or have at least one other user test this).

I'm back to my regular school schedule so I don't have much time, but I've started writing  a PAM module that authenticates you through a key press. Didn't work with regreet during my initial attempt, though, so I'll have to invest some more time. Key presses are not ideal in any case though because we'll have to run regreet under xwayland for it to work. Unfortunately I can't think of another good way to "mock" authentication that doesn't use text input.

**Comment by fufexan** [comment]
I've tested this PR with Howdy (with the aforementioned workaround disabled) and it works well. As soon as my face was recognized, I saw "Logging in..." and my session was started without any other button presses.

**Comment by rharish101** [comment]
> I've tested this PR with Howdy (with the aforementioned workaround disabled) and it works well. As soon as my face was recognized, I saw "Logging in..." and my session was started without any other button presses.

Great, then I'll work on the error message UI and then merge this PR.

FYI I also started the async port (without the `HandleGreetdResponse` stuff, i.e. with the `loop` inside `handle_greetd_response`), but I'm facing the previous issue of the GUI not being updated (contrary to what I thought). I might have to ask the Relm4 devs on Matrix whether `update` and `update_view` can truly run in parallel without waiting for the other.

**Comment by JohnnyJayJay** [comment]
> > I've tested this PR with Howdy (with the aforementioned workaround disabled) and it works well. As soon as my face was recognized, I saw "Logging in..." and my session was started without any other button presses.
> 
> Great, then I'll work on the error message UI and then merge this PR.
> 
> FYI I also started the async port (without the `HandleGreetdResponse` stuff, i.e. with the `loop` inside `handle_greetd_response`), but I'm facing the previous issue of the GUI not being updated (contrary to what I thought). I might have to ask the Relm4 devs on Matrix whether `update` and `update_view` can run in parallel (asynchronously, i.e. one runs when the other awaits) without waiting for the other.

We should use async components so we can make `handle_greetd_response`, the `update` function, and, most importantly, the greetd client async. And then `await` greetd responses.

**Comment by rharish101** [comment]
> We should use async components so we can make `handle_greetd_response`, the `update` function, and, most importantly, the greetd client async. And then `await` greetd responses.

Yup, I've already done all of that. The issue is using the previous looping approach (even in async) instead of sending out a `HandleGreetdResponse` command.

**Comment by JohnnyJayJay** [comment]
ok, yeah, that's weird. It *should* work with await in a loop, from what I understand. 

At any rate, that change should probably move to a new PR after merging this one, right?

**Comment by rharish101** [comment]
> At any rate, that change should probably move to a new PR after merging this one, right?

Of course. I just have to style the error messages now.

**Comment by JohnnyJayJay** [comment]
Hmm, I think having a separation of message and error makes sense because they are two different message types and both can be received at the same time. There are some problems with the idea of using the same place for errors and messages, but handling them differently. Here's an example from pam_fprintd:

1. message: Place your index finger on the fingerprint reader
2. I place something else
3. error: Failed to match fingerprint
4. message: Place your index finger on the fingerprint reader

This is already less than ideal because one of those messages is inevitably "skipped". And "switching" from an error to a previous message after a timeout additionally makes it seem like the message is "up to date" even though it may not be. Now, it's not possible to really get to the root of this problem, but if the message just *stays* somewhere without changing, I think it communicates the "timeline" better.

As for layouting problems, the way it is right now it looks the exact same as before to me (if the error isn't displayed). And when it's displayed the size of the box changes, that doesn't bother me (everything is still centered), but I suppose that's what you're referring to? I'm sure we could fix that and put it in its own box or something.

**Comment by rharish101** [comment]
I've made ReGreet display the errors like a notification that's shown at the bottom. Could you test it out to see if it's okay (UI/UX-wise)?

**Comment by fufexan** [comment]
I've tried this again, but the UI gets stuck after Howdy fails to authenticate me. Here's the log: [log.txt](https://github.com/rharish101/ReGreet/files/11009551/log.txt)

**Comment by rharish101** [comment]
Sadly, the logs weren't showing anything about the UI locking up. Could you try it again with the latest commit on this PR?

**Comment by fufexan** [comment]
Sure.

**Comment by fufexan** [comment]
[log.txt](https://github.com/rharish101/ReGreet/files/11009646/log.txt)

**Comment by rharish101** [comment]
You seem to have posted the exact same log file again 😄.

**Comment by fufexan** [comment]
I guess the log rotation doesn't work then. I will remove `/var/log/regreet/log` and try again.

**Comment by rharish101** [comment]
That's odd; the latest logs should always be appended to the latest file (`/var/log/regreet/log`). I can see it happening on my machine. Did you run it with `-l debug`? By default ReGreet uses a log level of `warn`, which should only log warnings and errors.

**Comment by fufexan** [comment]
Yes. This is the relevant part of the sway config that launches ReGreet:
`exec "/nix/store/v97dhaj52xg129826qv2py0l1z30c4n9-regreet-unstable-2023-03-18/bin/regreet -l debug; swaymsg exit"`.
The version is also fine, which means I'm using the new commit, and not the default package.

Anyway, here's the new log, which I don't see having anything new [log.txt](https://github.com/rharish101/ReGreet/files/11009685/log.txt).

**Comment by rharish101** [comment]
I see some panics in both the logs: maybe they're the actual cause? However, that should cause ReGreet to crash immediately instead of freezing. Secondly, the panic seems to be that ReGreet is unable to communicate with greetd to create a session. Does this only happen when Howdy fails to authenticate you, or does it happen all the time?

**Comment by fufexan** [comment]
When howdy recognizes me, it works fine. I'll try temporarily disabling Howdy and trying to input a wrong password.

**Comment by fufexan** [comment]
With howdy disabled, I get the proper error as shown in your screenshot.

**Comment by fufexan** [comment]
Ok, I've found out more. Greetd is the one actually crashing. Here's the log from `journalctl` [greetdlog.txt](https://github.com/rharish101/ReGreet/files/11009745/greetdlog.txt)

**Comment by rharish101** [comment]
Okay this is very weird. Have you tried to see if this happens with the `main` branch? If that one works but this one doesn't, then I think you might have to go through the trouble of bisecting the commit where things broke in this PR 😅.

I wish I could test this locally, but I don't think my laptop has an IR sensor for Howdy.

**Comment by rharish101** [comment]
I guess if the bug is with greetd, could you report it upstream? Try testing it with other greeters like gtkgreet to rule out ReGreet. If it's indeed a greetd issue, then we can go forward with this PR if the UI/UX of the error message is suitable.

Meanwhile I'm gonna have to learn to read core dumps 😅.

**Comment by fufexan** [comment]
Opened a bug report [here](https://todo.sr.ht/~kennylevinsen/greetd/41).

**Comment by JohnnyJayJay** [comment]
> I've made ReGreet display the errors like a notification that's shown at the bottom. Could you test it out to see if it's okay (UI/UX-wise)?
> 
> EDIT: Here's a screenshot: 
> 
> The error message pops up in an info bar with an animation, like a notification, before being hidden after a few seconds.

LGTM

**Comment by rharish101** [comment]
Merged!

**Comment by rharish101** [comment]
For posteriority:

> > FYI I also started the async port (without the HandleGreetdResponse stuff, i.e. with the loop inside handle_greetd_response), but I'm facing the previous issue of the GUI not being updated (contrary to what I thought). I might have to ask the Relm4 devs on Matrix whether update and update_view can run in parallel (asynchronously, i.e. one runs when the other awaits) without waiting for the other.
>
> ok, yeah, that's weird. It _should_ work with await in a loop, from what I understand.

Well, I talked to the Relm4 guys on Matrix, and they said that we still need async commands when running async. The only benefit we get is that a long-running async update function won't block the main thread. Hence, I went ahead and async-ed it, but we still use commands.

**Pull request merged**
